modules/shared/src/main/scala/org/tessellation/currency/dataApplication/package.scala

```diff
@@ -1,33 +1,150 @@
+import org.tessellation.schema.balance.Amount
 import org.tessellation.schema.round.RoundId
 import org.tessellation.schema.{GlobalIncrementalSnapshot, SnapshotOrdinal}
+import org.tessellation.security._
 import org.tessellation.security.hash.Hash
 import org.tessellation.security.signature.Signed
-import org.tessellation.security.{Encodable, Hashed, SecurityProvider}
 
+import derevo.cats.{order, show}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
 import eu.timepit.refined.auto._
 import io.circe._
 import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
 import io.circe.syntax._
 import org.http4s._
-import org.http4s.circe.jsonEncoderOf
+import org.http4s.circe.{jsonEncoderOf, jsonOf}
 import org.http4s.server.Router
 
-trait DataUpdate
+sealed trait DataTransaction
+object DataTransaction {
+  type DataTransactions = NonEmptyList[Signed[DataTransaction]]
+
+  implicit def encoder(implicit e: Encoder[DataUpdate]): Encoder[DataTransaction] =
+    Encoder.instance {
+      case v: FeeTransaction => v.asJson
+      case v: DataUpdate     => e.apply(v)
+    }
+
+  implicit def decoder(implicit d: Decoder[DataUpdate]): Decoder[DataTransaction] =
+    (c: HCursor) => Decoder[FeeTransaction].apply(c).orElse(d.apply(c))
+
+  private def collectTransactions[A <: DataTransaction](dataTransactions: List[DataTransactions])(
+    pf: PartialFunction[DataTransaction, A]
+  ): List[Signed[A]] =
+    dataTransactions.flatMap { currentDataTransactions =>
+      currentDataTransactions.collect { dataTransaction =>
+        dataTransaction.value match {
+          case transaction if pf.isDefinedAt(transaction) =>
+            Signed(pf(transaction), dataTransaction.proofs)
+        }
+      }
+    }
+
+  def getDataUpdates(dataTransactions: List[DataTransactions]): List[Signed[DataUpdate]] =
+    collectTransactions(dataTransactions) {
+      case dataUpdate: DataUpdate => dataUpdate
+    }
+
+  def getFeeTransactions(dataTransactions: List[DataTransactions]): List[Signed[FeeTransaction]] =
+    collectTransactions(dataTransactions) {
+      case feeTransaction: FeeTransaction => feeTransaction
+    }
+
+  def getDataTransactionHashes[F[_]: Async: Hasher](
+    data: List[DataTransactions],
+    serializeDataUpdate: DataUpdate => F[Array[Byte]]
+  ): F[List[NonEmptyList[hash.Hash]]] =
+    data.traverse { dataTransactions =>
+      dataTransactions.toList.traverse { signedTransaction =>
+        signedTransaction.value match {
+          case dataUpdate: DataUpdate =>
+            Signed(dataUpdate, signedTransaction.proofs)
+              .toHashed(serializeDataUpdate)
+              .map(_.hash)
+          case feeTransaction: FeeTransaction =>
+            Signed(feeTransaction, signedTransaction.proofs)
+              .toHashed(serialize[F])
+              .map(_.hash)
+        }
+      }.map(NonEmptyList.fromListUnsafe)
+    }
+
+  def getFeeTransactionOfDataUpdate[F[_]: Async](
```

Comment:
it should be defined either in `FeeTransaction`: `getByDataUpdate` or in `DataUpdate`: `getMatchingFeeTransaction`. probably the first one is better

Reply:
👍🏼